sound/soc/sof/core.c

```diff
@@ -135,6 +135,53 @@ EXPORT_SYMBOL(snd_sof_get_status);
  *	(System Suspend/Runtime Suspend)
  */
 
+static int sof_load_and_run_firmware(struct snd_sof_dev *sdev)
+{
+	/* load the firmware */
+	int ret = snd_sof_load_firmware(sdev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to load DSP firmware %d\n",
+			ret);
+		return ret;
+	}
+
+	sdev->fw_state = SOF_FW_BOOT_IN_PROGRESS;
+
+	/*
+	 * Boot the firmware. The FW boot status will be modified
+	 * in snd_sof_run_firmware() depending on the outcome.
+	 */
+	ret = snd_sof_run_firmware(sdev);
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to boot DSP firmware %d\n",
+			ret);
+		goto fw_run_err;
```

Comment:
Curious to know the usage of goto here.

I think we can avoid using it by directly unloading firmware and returning from here itself. If there was another situation in the function which would have used the same set of steps for cleanup purposes, then it would have been logical to write a goto for those common statements and use it (to avoid repeating).

Reply:
@ymdatta I think it's more or less a matter of choice here. If no resource freeing had to be done, i.e. if `snd_sof_fw_unload()` didn't have to be called, sure, we'd just return here. But since some clean up must be done, I actually decided to place it at the end of the function.